File: src/services/nodeAtPosition.test.ts
import { test } from "node:test";

import { expect } from "expect";

import { getIdentifierAtPosition, getNodeAtPosition } from "./nodeAtPosition.ts";
import { parseSourceFile } from "../compiler/parser.ts";
import { type Identifier, SyntaxKind } from "../compiler/types.ts";

const source = "class C {\n  int field;\n  void m() { int local = 1; }\n}\n";
const sf = parseSourceFile("T.java", source);

test("getNodeAtPosition finds the deepest node at an offset", () => {
  const node = getNodeAtPosition(sf, source.indexOf("field"));
  expect(node.kind).toBe(SyntaxKind.Identifier);
  expect((node as Identifier).text).toBe("field");
});

test("getIdentifierAtPosition resolves a name under the cursor", () => {
  const localOffset = source.indexOf("local");
  expect((getIdentifierAtPosition(sf, localOffset) as Identifier).text).toBe("local");
});

test("getIdentifierAtPosition accepts the trailing edge of a name", () => {
  const end = source.indexOf("field") + "field".length;
  expect((getIdentifierAtPosition(sf, end) as Identifier).text).toBe("field");
});

test("getIdentifierAtPosition returns undefined off any name", () => {
  // offset on the '{' of the class body
  expect(getIdentifierAtPosition(sf, source.indexOf("{"))).toBeUndefined();
});

File: togo/internal/compiler/node_at_position_test.go
package compiler

import (
	"strings"
	"testing"
)

// Port of src/services/nodeAtPosition.test.ts.

const napSource = "class C {\n  int field;\n  void m() { int local = 1; }\n}\n"

func napFile() *Node { return ParseSourceFile("T.java", napSource) }

func TestGetNodeAtPositionDeepest(t *testing.T) {
	node := GetNodeAtPosition(napFile(), strings.Index(napSource, "field"))
	if node.Kind != Identifier || node.AsIdentifier().Text != "field" {
		t.Errorf("got kind %v text %q, want Identifier 'field'", node.Kind, nodeText(node))
	}
}

func TestGetIdentifierAtPositionResolves(t *testing.T) {
	id := GetIdentifierAtPosition(napFile(), strings.Index(napSource, "local"))
	if id == nil || id.AsIdentifier().Text != "local" {
		t.Errorf("got %v, want 'local'", id)
	}
}

func TestGetIdentifierAtPositionTrailingEdge(t *testing.T) {
	end := strings.Index(napSource, "field") + len("field")
	id := GetIdentifierAtPosition(napFile(), end)
	if id == nil || id.AsIdentifier().Text != "field" {
		t.Errorf("got %v, want 'field' at trailing edge", id)
	}
}

func TestGetIdentifierAtPositionOffName(t *testing.T) {
	if id := GetIdentifierAtPosition(napFile(), strings.Index(napSource, "{")); id != nil {
		t.Errorf("got %v, want nil off any name", id)
	}
}

func nodeText(n *Node) string {
	if n.Kind == Identifier {
		return n.AsIdentifier().Text
	}
	return ""
}
